sql/core/src/test/scala/org/apache/spark/sql/execution/debug/DebuggingSuite.scala

```diff
@@ -48,4 +50,49 @@ class DebuggingSuite extends SparkFunSuite with SharedSQLContext {
     assert(res.forall{ case (subtree, code) =>
       subtree.contains("Range") && code.contains("Object[]")})
   }
+
+  test("SPARK-28537: DebugExec cannot debug broadcast or columnar related queries") {
+    val rightDF = spark.range(10)
+    val leftDF = spark.range(10)
+    val joinedDF = leftDF.join(rightDF, leftDF("id") === rightDF("id"))
+    try {
+      val captured = new ByteArrayOutputStream()
+      Console.withOut(captured) {
+        joinedDF.debug()
+      }
+
+      val output = captured.toString()
+      assert(output.contains(
+        """== BroadcastExchange HashedRelationBroadcastMode(List(input[0, bigint, false])) ==
+        |Tuples output: 0
+        | id LongType: {}
+        |== WholeStageCodegen ==
+        |Tuples output: 10
+        | id LongType: {java.lang.Long}
+        |== Range (0, 10, step=1, splits=2) ==
+        |Tuples output: 0
+        | id LongType: {}""".stripMargin))
+    } catch {
+      case e: Throwable => fail("debug() for broadcast failed with exception", e)
+    }
+
+    val df = spark.range(5)
+    df.persist()
+    try {
+      val captured = new ByteArrayOutputStream()
+      Console.withOut(captured) {
+        df.debug()
+      }
+
+      val exprId = df.queryExecution.executedPlan.output.head.toString
+      val output = captured.toString()
+      assert(output.contains(
```

Comment:
For these kinds of tests, substring matching seems to be better than exact matching.

Reply:
I'll follow the manner in [ExplainSuite](https://github.com/apache/spark/blob/e3418649dcb50f2a2fb977560d87a94c81516198/sql/core/src/test/scala/org/apache/spark/sql/ExplainSuite.scala#L35).